### PR TITLE
Fix/product review include variants

### PR DESCRIPTION
**Pull request by aragon999**
### 1. Why is this change necessary?
Currently if the product has variants and an assigned shopping experience layout only the reviews of that variant are displayed.

### 2. What does this change do, exactly?
Use the loading mechanism of the `ProductReviewLoader` to load the reviews. 

This PR is based on https://github.com/shopware/platform/pull/2387 since the widget route loads the template for the normal product page.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-19689

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2389"></a>

### CODE REVIEW

File: src/Core/Content/Product/Cms/ProductDescriptionReviewsCmsElementResolver.php [hunk 1]
```diff
@@ -7,34 +7,19 @@
...
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Aggregation\Bucket\TermsAggregation;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\AggregationResult\Bucket\TermsResult;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\EntitySearchResult;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\MultiFilter;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\RangeFilter;
-use Shopware\Core\Framework\DataAbstractionLayer\Search\Sorting\FieldSorting;
 use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Shopware\Storefront\Page\Product\Review\ProductReviewLoader;
```
Comment (mitelg) [inline src/Core/Content/Product/Cms/ProductDescriptionReviewsCmsElementResolver.php]: core classes are not allowed to have dependencies on classes from the Storefront package! this must work without the Storefront being installed

Comment (mitelg) [reply src/Core/Content/Product/Cms/ProductDescriptionReviewsCmsElementResolver.php]: you may want to use this class in the Storefront instead, to avoid duplicate code. (not checked by me, if this is really possible 😁)

Comment (aragon999) [reply src/Core/Content/Product/Cms/ProductDescriptionReviewsCmsElementResolver.php]: Thank you for the feedback :-) I see, I did not notice that I implemented it like this, but I will look into this how it can be solved.

**Comment by shyim** [comment]
@aragon999 what is the state there

**Comment by aragon999** [comment]
@shyim I rebased onto the current trunk and (hopefully) fixed all issues. It should be ready now.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    /**
+     * load reviews for one product. The request must contain the productId
+     * otherwise MissingRequestParameterException is thrown
+     *
+     * @throws MissingRequestParameterException
+     * @throws \Shopware\Core\Framework\DataAbstractionLayer\Exception\InconsistentCriteriaIdsException
+     */
+    public function load(Request $request, SalesChannelContext $context): ProductReviewLoaderResult
```
Comment (keulinho) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: can you add a unit test for this one? 

and please also add small unit tests fir the result struct and event

File: src/Storefront/Page/Product/Review/ProductReviewsLoadedCompatibilitySubscriber.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use Shopware\Core\Framework\Feature;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Storefront\Framework\Page\StorefrontSearchResult;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * @internal
+ *
+ * @deprecated tag:v6.6.0 will be re<p></p>moved without replacement, since \Shopware\Storefront\Page\Product\Review\ProductReviewsLoadedEvent will be removed
```
Comment (keulinho) [inline src/Storefront/Page/Product/Review/ProductReviewsLoadedCompatibilitySubscriber.php]:
```suggestion
 * @deprecated tag:v6.6.0 - reason:remove-subscriber - will be removed without replacement, since \Shopware\Storefront\Page\Product\Review\ProductReviewsLoadedEvent will be removed
```

File: src/Storefront/Page/Product/Review/ProductReviewsLoadedCompatibilitySubscriber.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * @internal
+     */
+    public function __construct(private readonly EventDispatcher $eventDispatcher)
+    {
+    }
+
+    public static function getSubscribedEvents()
+    {
+        Feature::triggerDeprecationOrThrow(
```
Comment (keulinho) [inline src/Storefront/Page/Product/Review/ProductReviewsLoadedCompatibilitySubscriber.php]: we cannot trigger deprecation warnings here, as this method will be called on every kernel boot, same for the the listener, we cannot throw deprecation warnings whenever the not deprecated event is thrown, with the `reason:remove-subscriber` keyword in the deprecation annotation of this class you can silence the phpstan errors that all methods need to throw deprecation messages

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 2 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: I added the tests for the result struct and event.

Regarding the unit test of the loader itself, should every code branch be tested? If you have an example in mind with a comparable test, it would probably help, as I am not really sure how extensive this test should be.

Comment (keulinho) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: 100% coverage should always be the goal :see_no_evil: 

but from glancing at the code i think we should also add at least one integration test to ensure all the stuff is handled correctly as this class is mainly creating criterias for the repo, having a unit test that mocks the repo does not provide that much value

Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: @keulinho I now added an unit test for the `ProductReviewLoader`. Regarding the integration test, I checked and I am not sure if one wants to expose this functionality to the storefront API and add an integration test for this route, or if one should write an integration test for this "loader".

With this change the `ProductReviewRoute` is not used anymore in the core, but rather the loader is used everywhere, so I am not sure if one wants to change the route instead of adding this loader: https://github.com/shopware/platform/blob/174f160d1930f995e63516c7bae0282dafa1d668/src/Core/Content/Product/SalesChannel/Review/ProductReviewRoute.php#L31-L32
And I if so, I am not sure how to keep the backwards compatibility, properly for Store API routes.

**Comment by keulinho** [comment]
@shopware/st-inventory can you take a look here? :eyes:

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 2 above)
Comment (keulinho) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: Sorry when i was not clear here, what i meant was to add integration test specifically and directly for this loader, 
really cool and nice that you added unit tests for this class :+1: 

i just would like to see one basic happy path test that really adds a product with 2 reviews and checks that the sorting and aggregation really works as expected

So you would test this class directly with phpunit and not over the store api 

Is that clearer for you now?

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            ->getResult()
+        ;
+
+        return $customerReviews->first();
+    }
+
+    private function handlePointsAggregation(Request $request, Criteria $criteria, ?CustomerEntity $customer): void
+    {
+        $points = $request->get('points', []);
+        if (\is_array($points) && \count($points) > 0) {
```
Comment (buzyka) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:143]: The case when the request has some points is not covered by unit tests. So code in this condition statement is not tested at all. It would be great if you could add test for it.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            ->getResult()
+        ;
+
+        $reviewResult = ProductReviewLoaderResult::createFrom($reviews);
+        $reviewResult->setProductId($request->get('productId'));
+        $reviewResult->setParentId($request->get('parentId'));
+
+        $aggregation = $reviews->getAggregations()->get('ratingMatrix');
+        if ($aggregation instanceof TermsResult) {
+            $matrix = $aggregation->getBuckets();
```
Comment (buzyka) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:66]: As I see this line is not covered by unit tests. But the case with no empty `$matrix` is one of general so it would be great if you could cover this case. This value makes an influence on further execution of the program

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    private function createCriteria(Request $request, SalesChannelContext $context): Criteria
+    {
+        $limit = (int) $request->get('limit', self::LIMIT);
+        $page = (int) $request->get('p', self::DEFAULT_PAGE);
+        $offset = $limit * ($page - 1);
+
+        $sort = (string) $request->get('sort', self::DEFAULT_SORTING);
+        if (!\in_array($sort, self::ALLOWED_SORTINGS, true)) {
+            $sort = self::DEFAULT_SORTING;
```
Comment (buzyka) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:88]: The case when an incorrect sorting value is declared in the request is not covered by the test. It would be great if you could cover this case.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 5 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:143]: I added a test which tests this parameter, and validates by checking the criteria: https://github.com/shopware/platform/pull/2389/files#diff-6b11a3de3d75954fd76a26118499db1e25887b9e4183cd326793289c6617c19bR178-R265

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 6 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:66]: I [added an aggregation](https://github.com/shopware/platform/pull/2389/files#diff-6b11a3de3d75954fd76a26118499db1e25887b9e4183cd326793289c6617c19bR87-R90) and [validate the result](https://github.com/shopware/platform/pull/2389/files#diff-6b11a3de3d75954fd76a26118499db1e25887b9e4183cd326793289c6617c19bR109-R113), which is actually the logic contained in `RatingMatrix`.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 7 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:88]: I added a test which has an invalid sorting: https://github.com/shopware/platform/pull/2389/files#diff-6b11a3de3d75954fd76a26118499db1e25887b9e4183cd326793289c6617c19bR178
And another test, which sets some other parameters for the request: https://github.com/shopware/platform/pull/2389/files#diff-6b11a3de3d75954fd76a26118499db1e25887b9e4183cd326793289c6617c19bR267

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 2 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php:47]: So :-) Finally I added the integration test as well :-)

**Review by keulinho — APPROVED** [review]

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoaderResult.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+namespace Shopware\Core\Content\Product\SalesChannel\Review;
+
+use Shopware\Core\Content\Product\Aggregate\ProductReview\ProductReviewEntity;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\EntitySearchResult;
+use Shopware\Core\Framework\Log\Package;
+
+#[Package('inventory')]
+class ProductReviewLoaderResult extends EntitySearchResult
+{
+    protected string $productId;
```
Comment (buzyka) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoaderResult.php]: Declaration strict type for all protected properties that were relocated from `Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult` leads to BC changes
```
[BC] CHANGED: Type of property Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult#$parentId changed from having no type to string|null
[BC] CHANGED: Type of property Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult#$productId changed from having no type to string
[BC] CHANGED: Type of property Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult#$matrix changed from having no type to Shopware\Core\Content\Product\SalesChannel\Review\RatingMatrix
[BC] CHANGED: Type of property Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult#$customerReview changed from having no type to Shopware\Core\Content\Product\Aggregate\ProductReview\ProductReviewEntity|null
[BC] CHANGED: Type of property Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult#$totalReviews changed from having no type to int
```

I'd recommend just deprecating it before 6.6

```php
    /**
     * @deprecated tag:v6.6.0 - Will be strictly declared as type: string
     *
     * @var string
     */
    protected $productId;

    /**
     * @deprecated tag:v6.6.0 - Will be strictly declared as type: ?string
     *
     * @var string|null
     */
    protected $parentId;

    /**
     * @deprecated tag:v6.6.0 - Will be strictly declared as type: RatingMatrix
     *
     * @var RatingMatrix
     */
    protected $matrix;

    /**
     * @deprecated tag:v6.6.0 - Will be strictly declared as type: ?ProductReviewEntity
     *
     * @var ProductReviewEntity|null
     */
    protected $customerReview;

    /**
     * @deprecated tag:v6.6.0 - Will be strictly declared as type: int
     *
     * @var int
     */
    protected $totalReviews;

```

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+     * parentId otherwise a MissingRequestParameterException is thrown
+     *
+     * @throws MissingRequestParameterException
+     * @throws \Shopware\Core\Framework\DataAbstractionLayer\Exception\InconsistentCriteriaIdsException
+     */
+    public function load(Request $request, SalesChannelContext $context): ProductReviewLoaderResult
+    {
+        $productId = $request->get('parentId') ?? $request->get('productId');
+        if (!$productId) {
+            throw new MissingRequestParameterException('productId');
```
Comment (buzyka) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php]: There is implemented new [domain exception schema](https://github.com/shopware/platform/blob/trunk/coding-guidelines/core/domain-exceptions.md) in the platform. If you could rebase on a fresh `trunk` this line brings to error.
You could create a separate static method for generating productId exception in `Shopware\Core\Content\Product\ProductException` 

```php
public static function missingProductId(string $path = ''): HttpException
    {
        if (!Feature::isActive('v6.6.0.0')) {
            return new MissingRequestParameterException('productId', $path);
        }

        return new self(
            Response::HTTP_BAD_REQUEST,
            self::PRODUCT_MISSING_PRODUCT_ID_IN_REQUEST_CODE,
            'Parameter "productId" is missing.',
            ['path' => $path]
        );
    }
```

and  call it here

`throw ProductException::missingProductId();`

**Comment by buzyka** [comment]
I've been working on integrating the changes into our current platform and encountered a few problems that need attention. I've provided detailed comments in the PR regarding the backward compatibility breaks, so please take a look.

Also, I noticed a few instances where the old class is still being used (ex [ProductController.php](https://github.com/shopware/platform/blob/trunk/src/Storefront/Controller/ProductController.php),  [ProductPageLoader](https://github.com/shopware/platform/blob/trunk/src/Storefront/Page/Product/ProductPageLoader.php), etc. ).  This might have been the reason why the other problems didn't surface earlier. It would be beneficial to begin refactoring by addressing these areas.

In my opinion, the root cause of these issues lies in the reversed inheritance scheme that has been implemented (the old class inherits from the new one). It is leading to the aforementioned backward compatibility problems and usage complications. While this may not pose a problem for internal classes, it becomes problematic for public classes. Take the `src/Storefront/Page/Product/Review/ProductReviewsWidgetLoadedHook.php` as an example.

To tackle this, it might need to consider adjusting the solution's architecture. One possible approach that comes to mind is modifying the inheritance scheme.

**Review by buzyka — CHANGES_REQUESTED** [review]
I've been working on integrating the changes into our current platform and encountered a few problems that need attention. I've provided detailed comments in the PR regarding the backward compatibility breaks, so please take a look.

Also, I noticed a few instances where the old class is still being used (ex [ProductController.php](https://github.com/shopware/platform/blob/trunk/src/Storefront/Controller/ProductController.php),  [ProductPageLoader](https://github.com/shopware/platform/blob/trunk/src/Storefront/Page/Product/ProductPageLoader.php), etc. ).  This might have been the reason why the other problems didn't surface earlier. It would be beneficial to begin refactoring by addressing these areas.

In my opinion, the root cause of these issues lies in the reversed inheritance scheme that has been implemented (the old class inherits from the new one). It is leading to the aforementioned backward compatibility problems and usage complications. While this may not pose a problem for internal classes, it becomes problematic for public classes. Take the `src/Storefront/Page/Product/Review/ProductReviewsWidgetLoadedHook.php` as an example.

To tackle this, it might need to consider adjusting the solution's architecture. One possible approach that comes to mind is modifying the inheritance scheme.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoaderResult.php (hunk 8 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoaderResult.php]: You are right, I implemented the changes as requested, such the BCs should be fixed.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php (hunk 9 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewLoader.php]: I rebased and changed to the new domain exceptions. ~~I think a new method is not required, as it is convenient to have the missing parameter as argument.~~

Edit: I first misunderstood the domain exceptions, I thought if there is something wrong with the request, than it is still in the "routing" domain :-) But I now implemented as suggested.

**Comment by aragon999** [comment]
I implemented the changes as requested. I am not sure if I understood the arguments correct, but I think the BC should be fixed. 

For the other remaining occurrences of the "old" `Shopware\Storefront\Page\Product\Review\ProductReviewLoader` it would be a breaking change, as the only remaining functionality of this class is to convert the core `ReviewLoaderResult` to the `Shopware\Storefront\Page\Product\Review\ReviewLoaderResult`.

Please let me know, if there are missing changes.

**Review by buzyka — CHANGES_REQUESTED** [review]
Sorry for a late response

Thank you very much for your update. I apologize that my previous comment was not clear enough. Below I've tried to explain it more clearly.

**_Problem:_**

In general deprecated classes can not be used anymore in core code because the shopware platform trying to be prepared for the next major version. Each PR checked for that. So we have a problem with this in this case. 
To see it you could set the environment variable  ` FEATURE_ALL="major"` and run tests. 
for example  `FEATURE_ALL="major"  php vendor/bin/phpunit --configuration phpunit.xml.dist src/Storefront/Test/Controller/ProductControllerTest.php`.

**_Just like a variant of a solution._**

 As one of the possible solution, I've proposed an inheritance scheme: 
`Shopware\Core\Content\Product\SalesChannel\Review `extends `Shopware\Storefront\Page\Product\Review\ProductReviewLoader`.  
In this case Shopware\Core\Content\Product\SalesChannel\Review extends the old class,  compatibility in safity and we could change everywhere in the platform, the old loader to a new one.

I hope it was clear enough. If you have any questions just let me know.

**Comment by aragon999** [comment]
> `Shopware\Core\Content\Product\SalesChannel\Review` extends `Shopware\Storefront\Page\Product\Review\ProductReviewLoader`

I think that is not possible, as the `Core` may not depend on the `Storefront`.

On the other hand, if we would replace the `Shopware\Storefront\Page\Product\Review\ProductReviewLoader` for example in the `Shopware\Storefront\Controller\ProductController` we need to somehow ensure that for Shopware 6.5 a `Shopware\Storefront\Page\Product\Review\ReviewLoaderResult` and not a `Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult` is returned, which is only possible by using a class from the `Storefront` package. But this is only needed for Shopware 6.5 and hence this class should be deprecated.

So in other words, I am not sure how to solve this without using a class which is deprecated for Shopware 6.5.

**Comment by buzyka** [comment]
> > `Shopware\Core\Content\Product\SalesChannel\Review` extends `Shopware\Storefront\Page\Product\Review\ProductReviewLoader`
> 
> I think that is not possible, as the `Core` may not depend on the `Storefront`.

I agree, core classes shouldn't be extended from Srotorefront.

> 
> On the other hand, if we would replace the `Shopware\Storefront\Page\Product\Review\ProductReviewLoader` for example in the `Shopware\Storefront\Controller\ProductController` we need to somehow ensure that for Shopware 6.5 a `Shopware\Storefront\Page\Product\Review\ReviewLoaderResult` and not a `Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult` is returned, which is only possible by using a class from the `Storefront` package. But this is only needed for Shopware 6.5 and hence this class should be deprecated.


As I  see currently class  `Shopware\Storefront\Page\Product\Review\ProductReviewLoader` 

As I  see currently class  `Shopware\Storefront\Page\Product\Review\ProductReviewLoader` used mostly in classes in which constructors are declared as internal,
` Shopware\Storefront\Controller\CmsController`
`Shopware\Storefront\Controller\ProductController`
  so theoretically it is possible to change them to a new class without breaking changes. 
  
The bigger problem with `Shopware\Core\Content\Product\SalesChannel\Review\ProductReviewResult`.
Instead `Shopware\Storefront\Page\Product\Review\ProductReviewWidgetLoadedHook` maybe it makes sense create a new hook wich will use `ProductReviewResult`. For `Shopware\Storefront\Page\Product\ProductPage` it might be possible to use union type for getter and setter. (Just one of variants).

It is a really difficult thing and requires big rework,  but we should find some solution to have a chance to merge PR into the platform.

**Comment by aragon999** [comment]
I will look into it and try to solve it, but probably not this week, as I am quite busy.

**Comment by buzyka** [comment]
Unfortunately, there have been no actions in this PR for a long time. I'll close it. If you find time to continue working on PR, feel free to open it again.
